text/0000-io-safety.md

```diff
@@ -0,0 +1,428 @@
+immediately; adoption could be gradual:
+
+ - First, `std` adds the new types and traits with impls for all the relevant
+   `std` types. This is a backwards-compatible change.
+
+ - After that, crates could begin to use the new types and implement the new
+   traits for their own types. These changes would be small and semver-compatible,
+   without special coordination.
+
+ - Once the standard library and enough popular crates implement the new
+   traits, crates could start to switch to using the new traits as bounds when
+   accepting generic arguments, at their own pace. These would be
+   semver-incompatible changes, though most users of APIs switching to these
+   new traits wouldn't need any changes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The I/O safety concept
+
+In addition to the Rust language's memory safety, Rust's standard library also
+guarantees I/O safety. An I/O operation is *valid* if the raw handles
+([`RawFd`], [`RawHandle`], and [`RawSocket`]) it operates on are values
+explicitly returned from the OS, and the operation occurs within the lifetime
+the OS associates with them. Rust code has *I/O safety* if it's not possible
+for that code to cause invalid I/O operations.
+
+While some OS's document their file descriptor allocation algorithms, a handle
+value predicted with knowledge of these algorithms isn't considered "explicitly
+returned from the OS".
+
+Functions accepting arbitrary raw I/O handle values ([`RawFd`], [`RawHandle`],
+or [`RawSocket`]) should be `unsafe` if they can lead to any I/O being
+performed on those handles through safe APIs.
+
+## `OwnedFd` and `BorrowedFd<'owned>`
+
+`OwnedFd` and `BorrowedFd` are both `repr(transparent)` with a `RawFd` value
+on the inside, and both can use niche optimizations so that `Option<OwnedFd>`
+and `Option<BorrowedFd<'_>>` are the same size, and can be used in FFI
+declarations for functions like `open`, `read`, `write`, `close`, and so on.
+When used this way, they ensure I/O safety all the way out to the FFI boundary.
+
+These types also implement the existing `AsRawFd`, `IntoRawFd`, and `FromRawFd`
+traits, so they can interoperate with existing code that works with `RawFd`
+types.
+
+## `AsFd`, `IntoFd`, and `FromFd`
+
+These types provide `as_fd`, `into_fd`, and `from_fd` functions similar to
+their `Raw` counterparts, but with the benefit of a safe interface, it's safe
+to provide a few simple conveniences which make the API much more flexible:
+
+ - A `from_into_fd` function which takes a `IntoFd` and converts it into a
+   `FromFd`, allowing users to perform this common sequence in a single
+   step, and without having to use `unsafe`.
+
+ - A `as_filelike_view::<T>()` function returns a `View`, which contains a
+   temporary `ManuallyDrop` instance of T constructed from the contained
+   file descriptor, allowing users to "view" a raw file descriptor as a
+   `File`, `TcpStream`, and so on.
```

Comment:
Do you propose to add this to the standard library, or is this just mentioned as a possible use of these APIs?

Reply:
I propose adding this to the standard library. Experience with it in unsafe-io and feedback in this RFC is that it's a relatively common thing to want to do with file descriptors, and it's a safe API that [requires `unsafe`](https://github.com/sunfishcode/io-experiment/blob/main/src/views.rs#L54) to implement.

That said, it isn't essential, and could also be provided by an extension trait in a crate.